web/src/hash_util.ts

```diff
@@ -193,6 +193,46 @@ export function parse_narrow(hash: string[]): Term[] | undefined {
     return terms;
 }
 
+/**
+ *
+ * @param url_str string containing a valid narrow url which can be transformed to
+ * `#**stream>topic**` syntax
+ * @returns `string[]` containing only the stream and topic (if present) in this order
+ * or `null` if the url is not a valid narrow url
+ */
+export function decode_stream_topic_from_url(url_str: string): string[] | null {
+    try {
+        const index = url_str.indexOf("#narrow");
+        const origin = new URL(url_str).origin;
+        if (origin !== window.location.origin) {
+            return null;
+        }
+        const terms = parse_narrow(url_str.slice(index).split(/\//));
+        if (terms === undefined) {
+            return null;
+        }
+        if (terms.length === 0 || terms.length > 2) {
+            // link should only contain stream and topic
+            // near/ links are not transformed
+            return null;
+        }
+        // this check is important as a malformed url
+        // may have `stream`, `topic` or `near:` in any order
+        if (terms[0].operator !== "stream") {
+            return null;
+        }
+        if (terms.length === 1) {
+            return [terms[0].operand];
+        }
+        if (terms[1].operator !== "topic") {
+            return null;
+        }
+        return [terms[0].operand, terms[1].operand];
```

Comment:
I'd still prefer this return an object with optional `stream`  and `topic` fields rather than a tuple.

Reply:
I've replaced the tuple with an object with `stream` and optional `topic` fields. 
I haven't kept `stream` optional and returning `null` in an invalid case as I think `null` makes it (logically) clear that the url provided was not suitable for transformation.

Reply:
OK